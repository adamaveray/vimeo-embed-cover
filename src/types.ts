export interface AspectRatio {
  width: number;
  height: number;
}

export type AdjustmentPercentage = number;

export interface Adjustments {
  width?: AdjustmentPercentage;
  left?: AdjustmentPercentage;
  height?: AdjustmentPercentage;
  top?: AdjustmentPercentage;
}

export interface Size {
  width: number;
  height: number;
}
