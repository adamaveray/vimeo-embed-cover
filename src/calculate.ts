import type { Adjustments, AspectRatio, Size } from './types';

export default function calculate(video: AspectRatio, container: Size): Adjustments {
  if (video.width <= 0 || video.height <= 0 || container.width <= 0 || container.height <= 0) {
    // Invalid values - cannot calculate
    return {};
  }

  const videoRatio = video.height / video.width;
  const containerRatio = container.height / container.width;

  /* eslint-disable no-case-declarations */
  switch (true) {
    case containerRatio > videoRatio:
      // Container too wide - stretch to fill
      const comparisonA = containerRatio / videoRatio;
      return {
        width: comparisonA,
        left: -(comparisonA - 1) / 2,
      };

    case videoRatio > containerRatio:
      // Container too tall - stretch to fill
      const comparisonB = videoRatio / containerRatio;
      return {
        height: comparisonB,
        top: -(comparisonB - 1) / 2,
      };

    default:
      // Perfect match
      return {};
  }
  /* eslint-enable no-case-declarations */
}
