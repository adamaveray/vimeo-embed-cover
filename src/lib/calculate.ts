import { type Adjustments, type AspectRatio, type Size } from '../types';

const HALF_MULTIPLIER = 0.5;

export default function calculate(video: AspectRatio, container: Size): Adjustments {
  if (video.width <= 0 || video.height <= 0 || container.width <= 0 || container.height <= 0) {
    // Invalid values - cannot calculate
    return {};
  }

  const videoRatio = video.height / video.width;
  const containerRatio = container.height / container.width;

  switch (true) {
    case containerRatio > videoRatio: {
      // Container too wide - stretch to fill
      const comparisonA = containerRatio / videoRatio;
      return {
        width: comparisonA,
        left: -(comparisonA - 1) * HALF_MULTIPLIER,
      };
    }

    case videoRatio > containerRatio: {
      // Container too tall - stretch to fill
      const comparisonB = videoRatio / containerRatio;
      return {
        height: comparisonB,
        top: -(comparisonB - 1) * HALF_MULTIPLIER,
      };
    }

    default: {
      // Perfect match
      return {};
    }
  }
}
