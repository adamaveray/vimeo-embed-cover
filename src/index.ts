import type { Player } from '@vimeo/player';
import type { AspectRatio } from './types';
import calculate from './calculate';

export * from './types';

const formatCssPercentage = (value: number): string => value * 100 + '%';

export default async function vimeoEmbedCover(player: Player): Promise<void> {
  await player.ready();

  const playerElement = (player as any).element as HTMLIFrameElement; // Not included in documented API but is available
  const containerElement = playerElement.parentElement;
  if (containerElement == null) {
    // Player element detached from DOM
    return;
  }

  const [videoWidth, videoHeight] = await Promise.all([player.getVideoWidth(), player.getVideoHeight()]);
  let aspectRatio: AspectRatio = { width: videoWidth, height: videoHeight };

  const update = (containerSize: DOMRect): void => {
    const adjustments = calculate(aspectRatio, containerSize);
    const { style } = playerElement;
    style.position = 'absolute';
    style.width = formatCssPercentage(adjustments.width ?? 1);
    style.height = formatCssPercentage(adjustments.height ?? 1);
    style.left = formatCssPercentage(adjustments.left ?? 0);
    style.top = formatCssPercentage(adjustments.top ?? 0);
  };

  const measureThenUpdate = (): void => {
    update(containerElement.getBoundingClientRect());
  };

  player.on('resize', (event) => {
    aspectRatio = { width: event.videoWidth, height: event.videoHeight };
    measureThenUpdate();
  });

  if (typeof ResizeObserver !== 'undefined') {
    // Update when wrapping element dimensions change
    new ResizeObserver((entries) => {
      update(((entries as any)[0] as ResizeObserverEntry).contentRect);
    }).observe(containerElement);
  } else {
    // Update when window changes on legacy browsers
    window.addEventListener('resize', measureThenUpdate);
  }

  // Set initial position
  measureThenUpdate();
}
