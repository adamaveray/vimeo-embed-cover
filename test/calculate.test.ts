/* eslint @typescript-eslint/no-magic-numbers: "off" -- Too many numbers to test */

import { type AspectRatio, type Size } from '../src';
import calculate from '../src/lib/calculate';

describe('the adjustments', () => {
  it.each<[string, AspectRatio]>([
    ['Same height, slightly wider', { width: 250, height: 100 }],
    ['Same height, significantly wider', { width: 1000, height: 100 }],
    ['Taller, slightly wider', { width: 600, height: 200 }],
  ])('are correct for too-wide videos', (label, aspectRatio) => {
    expect(calculate(aspectRatio, { width: 200, height: 100 })).toMatchSnapshot(label);
  });

  it.each<[string, AspectRatio]>([
    ['Same width, slightly taller', { width: 200, height: 150 }],
    ['Same width, significantly taller', { width: 200, height: 1000 }],
    ['Wider, slightly taller', { width: 400, height: 800 }],
  ])('are correct for too-tall videos', (label, aspectRatio) => {
    expect(calculate(aspectRatio, { width: 200, height: 100 })).toMatchSnapshot(label);
  });

  it.each<[string, AspectRatio]>([
    ['Identical', { width: 200, height: 100 }],
    ['Larger', { width: 400, height: 200 }],
    ['Smaller', { width: 100, height: 50 }],
  ])('are empty for same-ratio videos', (_, aspectRatio) => {
    expect(calculate(aspectRatio, { width: 200, height: 100 })).toEqual({});
  });

  it.each<AspectRatio>([
    { width: 10, height: 0 },
    { width: 0, height: 10 },
    { width: 0, height: 0 },
    { width: 10, height: -1 },
    { width: -1, height: 10 },
    { width: -1, height: -1 },
  ])('handle invalid video sizes', (aspectRatio) => {
    expect(calculate(aspectRatio, { width: 10, height: 10 })).toEqual({});
  });

  it.each<Size>([
    { width: 0, height: 10 },
    { width: 10, height: 0 },
    { width: 0, height: 0 },
    { width: -1, height: 10 },
    { width: 10, height: -1 },
    { width: -1, height: -1 },
  ])('handle invalid container sizes', (containerSize) => {
    expect(calculate({ width: 10, height: 10 }, containerSize)).toEqual({});
  });
});
