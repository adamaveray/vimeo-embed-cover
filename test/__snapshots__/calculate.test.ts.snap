// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the adjustments are correct for too-tall videos: Same width, significantly taller 1`] = `
{
  "height": 10,
  "top": -4.5,
}
`;

exports[`the adjustments are correct for too-tall videos: Same width, slightly taller 1`] = `
{
  "height": 1.5,
  "top": -0.25,
}
`;

exports[`the adjustments are correct for too-tall videos: Wider, slightly taller 1`] = `
{
  "height": 4,
  "top": -1.5,
}
`;

exports[`the adjustments are correct for too-wide videos: Same height, significantly wider 1`] = `
{
  "left": -2,
  "width": 5,
}
`;

exports[`the adjustments are correct for too-wide videos: Same height, slightly wider 1`] = `
{
  "left": -0.125,
  "width": 1.25,
}
`;

exports[`the adjustments are correct for too-wide videos: Taller, slightly wider 1`] = `
{
  "left": -0.25,
  "width": 1.5,
}
`;
